e2e/config/vhd.go

```diff
@@ -0,0 +1,176 @@
+	}
+	VHDAzureLinuxV2Gen2Arm64 = &VHD{
+		ImageID:         "/subscriptions/8ecadfc9-d1a3-4ea4-b844-0d9f87e4d7c8/resourceGroups/aksvhdtestbuildrg/providers/Microsoft.Compute/galleries/PackerSigGalleryEastUS/images/AzureLinuxV2Gen2Arm64",
+		VersionTagName:  "branch",
+		VersionTagValue: "refs/heads/master",
+	}
+	VHDAzureLinuxV2Gen2 = &VHD{
+		ImageID:         "/subscriptions/8ecadfc9-d1a3-4ea4-b844-0d9f87e4d7c8/resourceGroups/aksvhdtestbuildrg/providers/Microsoft.Compute/galleries/PackerSigGalleryEastUS/images/AzureLinuxV2Gen2",
+		VersionTagName:  "branch",
+		VersionTagValue: "refs/heads/master",
+	}
+	VHDCBLMarinerV2Gen2Arm64 = &VHD{
+		ImageID:         "/subscriptions/8ecadfc9-d1a3-4ea4-b844-0d9f87e4d7c8/resourceGroups/aksvhdtestbuildrg/providers/Microsoft.Compute/galleries/PackerSigGalleryEastUS/images/CBLMarinerV2Gen2Arm64",
+		VersionTagName:  "branch",
+		VersionTagValue: "refs/heads/master",
+	}
+	VHDCBLMarinerV2Gen2 = &VHD{
+		ImageID:         "/subscriptions/8ecadfc9-d1a3-4ea4-b844-0d9f87e4d7c8/resourceGroups/aksvhdtestbuildrg/providers/Microsoft.Compute/galleries/PackerSigGalleryEastUS/images/CBLMarinerV2Gen2",
+		VersionTagName:  "branch",
+		VersionTagValue: "refs/heads/master",
+	}
+)
+
+// VHDResourceID represents a resource ID pointing to a VHD in Azure. This could be theoretically
+// be the resource ID of a managed image or SIG image version, though for now this will always be a SIG image version.
+type VHDResourceID string
+
+func (id VHDResourceID) Short() string {
+	sep := "Microsoft.Compute/galleries/"
+	str := string(id)
+	if strings.Contains(str, sep) && !strings.HasSuffix(str, sep) {
+		return strings.Split(str, sep)[1]
+	}
+	return str
+}
+
+// This is ugly, but I wanted to achieve multiple things:
+// 1. If buildID specified use the image from the build, skip scenario if not found
+// 2. Fetch resource id only once (multiple scenarios can share the VHD)
+// 3. Call fetch method only if tests is scheduled to run (it takes times)
+func (v *VHD) ResourceID() VHDResourceID {
+	if v.BuildResourceID() != "" {
+		return v.BuildResourceID()
+	}
+	return v.NonBuildResourceID()
+}
+
+func (v *VHD) BuildResourceID() VHDResourceID {
+	v.fetchBuildResourceIDOnce.Do(func() {
+		v.buildResourceID, _ = findLatestImageWithTag(v.ImageID, "buildId", VHDBuildID)
+	})
+	return v.buildResourceID
+}
+
+func (v *VHD) NonBuildResourceID() VHDResourceID {
+	v.fetchResourceIDOnce.Do(func() {
+		if v.resourceID != "" {
+			return
+		}
+		v.resourceID, _ = findLatestImageWithTag(v.ImageID, v.VersionTagName, v.VersionTagValue)
+	})
+	return v.resourceID
+}
+
+var ErrNotFound = fmt.Errorf("not found")
+
+func findLatestImageWithTag(imageID, tagName, tagValue string) (VHDResourceID, error) {
+	ctx, cancel := context.WithTimeout(context.TODO(), time.Minute)
+	defer cancel()
+	image, err := parseImageID(imageID)
+	if err != nil {
+		return "", err
+	}
+
+	cred, err := azidentity.NewDefaultAzureCredential(nil)
```

Comment:
I guess this is the magic why we can get rid of the PAT. What pre-requisites do we need to get the cred from this function?

Reply:
This part hasn't been changed much. These credentials are used for every Azure API call. Like accessing SIG Images, creating AKS clusters etc.

The PAT token was used only to fetch artifacts from ADO. To extract a VHD resource id from it.